autorest/azure/azure.go

```diff
@@ -70,3 +92,33 @@ func ExtractClientID(resp *http.Response) string {
 func ExtractRequestID(resp *http.Response) string {
 	return autorest.ExtractHeaderValue(HeaderRequestID, resp)
 }
+
+// WithAzureErrorCheck handles the HTTP error responses sent by Azure service
+// response if the response conforms the Azure API error format. If there is a
+// chance service may return responses other than this format, the error is not
+// handled. In case the response is a proper Azure error, than the response body
+// is read and closed by this method and does not need further closing.
+func WithAzureErrorCheck() autorest.RespondDecorator {
+	return func(r autorest.Responder) autorest.Responder {
+		return autorest.ResponderFunc(func(resp *http.Response) error {
+			err := r.Respond(resp)
+			if err == nil && resp.StatusCode/100 >= 4 { // 4xx, 5xx
+				var e *AzureError
+
+				d := json.NewDecoder(resp.Body)
+				defer resp.Body.Close()
+				_ = d.Decode(&e)
+				// the reason we discard decoding error is we check for
+				// the Error field of AzureError anyway and in case of not
+				// returning any error here will be handled by the next responder.
+
+				if e != nil && e.ServiceError != nil {
+					e.RequestID = ExtractRequestID(resp)
+					e.StatusCode = resp.StatusCode
+					err = e
+				}
+			}
+			return err
```

Comment:
@ahmetalpbalkan I like the general idea, but believe we should tweak this a bit.

First, returning nil and closing the Body will cause problems for subsequent Responders. I suggest that either the code read / replace the Body (so subsequent Responders have something to work with) or that it always return an error.

Second, since this code is within the "azure" package, I see no need to include Azure in any of the names (e.g., azure.WithErrrorUnlessStatusCode, azure.ServiceError, and azure.Error seems sufficiently clear and stutter free). Also, I feel it best to move this code into azure/error.go and azure/error_test.go files.

Further, it may be better if the code followed the pattern of autorest.WithErrorUnlessStatusCode (or similar). Services document their error codes in the corresponding Swagger (and documentation), so the check can be precise. If you are uncomfortable with that, provide a default set of codes or default predicate to catch the error codes.

Lastly, I'd feel better the code manufactured an error if the json.Decode fails. The error could be generic and possibly populated with details from the Decode error which may be important under different debugging scenarios.

Ok?